source/framework/core/src/TRestDataSet.cxx

```diff
@@ -330,6 +349,12 @@ void TRestDataSet::GenerateDataSet() {
 
     fDataSet = MakeCut(fCut);
 
+    // Adding new user columns added to the dataset
+    for (const auto& [cName, cExpression] : fColumnNameExpressions) {
+        finalList.emplace_back(cName);
+        this->Define(cName, cExpression);
```

Comment:
I would change the behaviour of the `Define` function to something similar to `MakeCut` e.g:
```
 fDataSet = Define(cName, cExpression);
```
And in the `Define` function
```
ROOT::RDF::RNode TRestDataSet::Define(const std::string& columnName, const std::string& formula) {
  auto df = fDataSet;  
  std::string evalFormula = formula;
    for (auto const& [name, properties] : fQuantity)
        evalFormula =
            REST_StringHelper::Replace(evalFormula, name, DoubleToString(properties.value, "%12.10e"));
    df = df.Define(columnName, evalFormula);
    return df;
}
```
Then it should be possible to define a new observable but you should be capable to use or store de dataframe somewhere else. I think is more useful this way.

Reply:
Actually the `RDataFrame` is already being returned as it is done in `MakeCut` the only difference I see is that I just do not do `fDataSet = this->Define( ...)` since I know that `fDataSet` has been already updated inside.

Reply:
Also, I do not create a temporal `df` inside the method, it is this necessary?

Reply:
Oh what you want is that the internal `fDataSet` is not modified at all, then the returned contains the new column?

What if I want to add 3 new columns that use relevant quantities?

Reply:
Ok, what you want is to have the possibility to add a new column in a returned `RDataFrame`, but  this limits the operation to just 1-column, isn't? I mean, if you want to use the method outside the `dataset` you will only be able to use relevant quantities once.

Perhaps we should give as argument then the `data frame` where we will add the new evaluated column?

Reply:
I guess that becomes complicated

Reply:
Then I just change it to `AddColumn`

Reply:
Ok, finally I just leave it as `MakeCut` philosophy to be consistent.